### ISSUE TITLE
webpki-roots is copy-left

### ISSUE THREAD

**Issue by emilk** [algesten/ureq#478]
Thanks for a great library!

https://github.com/rustls/webpki-roots has a copy-left license (MPL-2.0) making it unsuitable for any closed-source use.

`webpki-roots` is enabled by the `tls` feature in `ureq (a default feature).

I suggest we make `rustls-native-certs` the default, and `webpki-roots` opt-in so that `ureq` does not create copy-left problems for its users unless they actively opt-in to using `webpki-roots`.

I can make a PR for this change.

---

PS: this was discovered using [`cargo-deny`](https://github.com/EmbarkStudios/cargo-deny)

**Comment by jsha** [comment]
Thanks for spotting this!

I'm a little reluctant to change the default, in part because https://github.com/rustls/rustls-native-certs/issues/25 means that the platform verifier trusts some untrusted roots, which creates risk of accepting an inappropriate certificate.

Also, this would definitely merit a major version bump, because it could break connecting to some sites that would otherwise work. Still, overall the native-certs approach is the one we'd like to take long-term, assuming rustls adds the necessary extension points.

What do you think of waiting on this change until rustls has the necessary extension points to use the platform verifier _and_ platform roots, then doing it as a major version bump?

**Comment by algesten** [comment]
I agree, @emilk we agree in principle to switch to `rustls-native-certs`, however now is not the time.

Would a way forward, for now, be to make a PR that documents the copyleft concerns?

**Comment by emilk** [comment]
I understand this is a big change not to be taken lightly, and I trust your judgement on this.

I really enjoy the minimal dependencies of `ureq` (and not having to pull in an async executor), so I hope this can be resolved in the near future!

Until then I'll be putting this in my `Cargo.toml`:

``` toml
[patch.crates-io]
ureq = { git = "https://github.com/emilk/ureq/", branch = "opt-in-webpki-roots" }
```

Thanks for your attention and responses, and for a great crate!

**Comment by emilk** [comment]
How about adding a new features to `ureq` in the meantime:

`tls_native_certs = ["webpki", "native-certs", "rustls"]`

This would allow users to use `tls` without any copy-left code being pulled in, by using

`ureq = { version = "2.4", default-features = false, features = ["gzip", "tls_native_certs"] }`

No major version bump required!

**Comment by dralley** [comment]
I clarified [here](https://github.com/algesten/ureq/pull/479#issuecomment-1031016683) that there should not be any "copyleft problems" due to including MPLv2 software such as `webpki-roots`.

**Comment by emilk** [comment]
Thanks for that clarification @dralley - that really makes this a non-issue as far as I can tell.

**Comment by emilk** [comment]
Closed, because MPL-2.0 is not really copy-left (see https://github.com/algesten/ureq/issues/478#issuecomment-1031029647)

**Comment by dralley** [comment]
It definitely is copyleft, but copyleft is unfairly associated with a couple of restrictions that are specific to the GNU licenses and don't apply to all copyleft licenses.
